### PR TITLE
[FLINK-16967][metrics][Slf4j] Refactor Slf4jReporter to be instantiat…

### PR DESCRIPTION
**Pull request by molsionmo**
…ed through Slf4jReporterFactory that implements MetricReporterFactory

## What is the purpose of the change

*Adds a new `MetricsReporterFactory` implementation to support loading `Slf4jReporter` through a factory based approach.*

## Brief change log

  - *Refactor Slf4jReporter to be instantiated through Slf4jReporterFactory that implements MetricReporterFactory*

## Verifying this change

This change is a trivial rework / code cleanup without any test coverage.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (no)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (no)
  - The serializers: (no)
  - The runtime per-record code paths (performance sensitive): (no)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: (no)
  - The S3 file system connector: (no)

## Documentation

  - Does this pull request introduce a new feature? (yes)
  - If yes, how is the feature documented? (JavaDocs)

### CODE REVIEW

File: flink-metrics/flink-metrics-slf4j/src/main/java/org/apache/flink/metrics/slf4j/Slf4jReporter.java [hunk 1]
```diff
@@ -39,6 +41,8 @@
 /**
  * {@link MetricReporter} that exports {@link Metric Metrics} via SLF4J {@link Logger}.
  */
+@PublicEvolving
```
Comment (zentol) [inline flink-metrics/flink-metrics-slf4j/src/main/java/org/apache/flink/metrics/slf4j/Slf4jReporter.java:44]: This shouldn't be added, as it is an unrelated change.

Comment (molsionmo) [reply flink-metrics/flink-metrics-slf4j/src/main/java/org/apache/flink/metrics/slf4j/Slf4jReporter.java:44]: ok, i will resubmit after delete it

Comment (molsionmo) [reply flink-metrics/flink-metrics-slf4j/src/main/java/org/apache/flink/metrics/slf4j/Slf4jReporter.java]: cc zentol this change have commit,please check it ,thx

File: flink-metrics/flink-metrics-slf4j/src/main/java/org/apache/flink/metrics/slf4j/Slf4jReporter.java [hunk 2]
```diff
@@ -18,6 +18,8 @@
 
 package org.apache.flink.metrics.slf4j;
 
+import org.apache.flink.annotation.PublicEvolving;
```
Comment (zentol) [inline flink-metrics/flink-metrics-slf4j/src/main/java/org/apache/flink/metrics/slf4j/Slf4jReporter.java:21]: this import is now unused

File: flink-end-to-end-tests/test-scripts/common.sh [hunk 3]
```diff
@@ -587,7 +587,7 @@ function kill_random_taskmanager {
 function setup_flink_slf4j_metric_reporter() {
   INTERVAL="${1:-1 SECONDS}"
   add_optional_lib "metrics-slf4j"
```
Comment (zentol) [inline flink-end-to-end-tests/test-scripts/common.sh:589]:
```suggestion
  add_optional_plugin "metrics-slf4j"
```

This is just a small formality to have the reporter be really loaded as a plugin.

Comment (molsionmo) [reply flink-end-to-end-tests/test-scripts/common.sh]: yes , you are right , i resumit after change it
